src/hclust.jl

```diff
@@ -531,6 +531,101 @@ function rorder!(hmer::HclustMerges)
     return hmer
 end
 
+
+"""
+    orderleaves_barjoseph!!(order::Vector{Int}, hcl::Hclust, dm::Array{Float64,2})
+
+Given a hierarchical cluster and the distance matrix used to generate it,
+use fast algorithm to determine optimal leaf order minimizing the distance
+between adjacent leaves. This is done using a heuristic where,
+when combining multi-leaf sub branches, only the outermost leaves are
+compared (a maximum of 4 comparisons per intersection).
+
+Sub branches are flipped if necessary to minimize the distance between adjacent
+nodes, and then the combined branches are treated as a block for future
+comparisons.
+
+Based on:
+[Bar-Joseph et. al. "Fast optimal leaf ordering for hierarchical clustering." _Bioinformatics_. (2001)](https://doi.org/10.1093/bioinformatics/17.suppl_1.S22)
+"""
+function orderleaves_barjoseph!(hcl::Hclust, dm::Array{Float64,2})
```

Comment:
```suggestion
function orderleaves_barjoseph!(hmer::HclustMerges, d::AbstractMatrix)
```
Note that `d` might be different from `Matrix{Float64}`.